### ISSUE TITLE
Do not update sincedb if the data is not accepted by elasticsearch

### ISSUE THREAD

**Issue by talevy** [elasticsearch/logstash#2540]
I am using elasticsearch-1.1.1 with logstash 1.4.0 and I am currently in process of parsing 1.2B documents across 18 indices.
At times, due to various reasons (probably a bug in LS), I see these errors:
```
{:timestamp=>"2014-05-13T18:32:06.403000+0000", :message=>"Failed to flush outgoing items", :outgoing_count=>2616, :exception=>#<RuntimeError: Non-OK response code from Elasticsearch: 404>, :backtrace=>["/elasticsearch/logstash/logstash-1.4.0/lib/logstash/outputs/elasticsearch/protocol.rb:127:in `bulk_ftw'", "/elasticsearch/logstash/logstash-1.4.0/lib/logstash/outputs/elasticsearch/protocol.rb:80:in `bulk'", "/elasticsearch/logstash/logstash-1.4.0/lib/logstash/outputs/elasticsearch.rb:331:in `flush'", "/elasticsearch/logstash/logstash-1.4.0/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/buffer.rb:219:in `buffer_flush'", "org/jruby/RubyHash.java:1339:in `each'", "/elasticsearch/logstash/logstash-1.4.0/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/buffer.rb:216:in `buffer_flush'", "/elasticsearch/logstash/logstash-1.4.0/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/buffer.rb:193:in `buffer_flush'", "/../logstash/logstash-1.4.0/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/buffer.rb:112:in `buffer_initia
```

The problem vanishes after I restart the logstash process.
Anyway, the major problem is that even though the write has failed, logstash updates sincedb and doesn't retry the loglines, this has caused a large pain as one has to constantly monitor and restart if such errors are found. So far I have lost over 100m documents and no idea how to get them back !

migrated from: https://logstash.jira.com/browse/LOGSTASH-2203

**Comment by jordansissel** [comment]
> and doesn't retry the loglines

I'm pretty sure it does retry, and if it doesn't in 1.4.2, this has been fixed to retry in Logstash 1.5.0 when released.

**Comment by jordansissel** [comment]
@talevy can you confirm?

**Comment by talevy** [comment]
@jordansissel I believe the problem here is when there may exist a network issue when contacting ES. It will retry forever, but if one restarts Logstash, they will lose their place because of the buffered elements. I am not sure how the file input does pre-fetching, and how many events that brings in before acknowledging in sincedb.

in general, I believe this issue is more interested in a safer logstash restart.

**Comment by jordansissel** [comment]
Ahh, two bugs here I suppose. I'll update with more details

**Comment by jordansissel** [comment]
I'll roughly summarize the report as I see it:

1) "Failed to flush outgoing items" occurs in Logstash and repeats forever, never resolving
2) A restart of Logstash resolves the "Failed to flush outgoing items" problem
3) Logstash will lose in-flight messages on restart
4) "100m documents" have been lost due to repeatedly restarting logstash.

"Retry forever" is intentional. If Logstash is having trouble talking to Elasticsearch, a restart of Logstash should *NOT* fix this.


> The problem vanishes after I restart the logstash process. 

The error reported "RuntimeError: Non-OK response code from Elasticsearch: 404" - I'm not exactly sure what caused this. It could be caused by aattempting to write to a closed index?
The problem is likely related to certain events that are causing Elasticsearch to fail the request. Restarting causes a loss of those events (see below on item 3) and allows logstash to continue on next start.

---

As for 3 (losing inflight messages). This is a known issue that will be solved by #1939.

As for 4 (100m documents lost), I don't really know what to say. In order for logstash to loose 100 _million_ documents across a restart, and assuming the default settings in Logstash, you'd have to restart Logstash at least 200 times (flush_size 5000 by default). If you need to restart logstash 200 times due to an error, I would suggest that continuing to apply percussive maintenance on Logstash will not result in a successful experience.

**Comment by jordansissel** [comment]
I can confirm that under Elasticsearch 1.1.1, if you attempt to write to a closed index with a bulk write will *totally* fail (404 on the request). Under Elasticsearch 1.4.1, this request error does not occur.

**Comment by jordansissel** [comment]
So!

**Comment by jordansissel** [comment]
This *specific* report of Logstash getting stuck. Workarounds which do *not* require restarts of Logstash are:
1) to reopen the old index(es) that are closed to allow Logstash to write and continue
2) to have Logstash drop any events that you know to be part of a closed index.

**Comment by jordansissel** [comment]
For the rest of the problem (logstash dropping in-flight messages on restart), will be solved by #1939
